ores/scoring_context.py

```diff
@@ -0,0 +1,224 @@
+
+        return model_info
+
+    def format_id_string(self, model_name, rev_id, injection_cache=None):
+        version = self.model_version(model_name)
+        score_id = ":".join(
+            str(v) for v in [self.name, model_name, version, rev_id])
+        if injection_cache is None:
+            return score_id
+        else:
+            sorted_tuple = tuple(sorted(injection_cache.items()))
+            cache_hash = sha1(bytes(str(sorted_tuple), 'utf8')).hexdigest()
+            return score_id + ":" + cache_hash
+
+    def _get_model_info_for(self, model_name):
+        return self[model_name].format_info(format='json')
+
+    def model_version(self, model_name):
+        return self[model_name].version
+
+    def model_features(self, model_name):
+        return self[model_name].features
+
+    def process_model_scores(self, model_names, root_cache, include_features):
+        model_scores = {}
+
+        for model_name in model_names:
+            model_scores[model_name] = {}
+
+            # Mostly CPU
+            model_scores[model_name]['score'] = \
+                self.process_score(model_name, dependency_cache=root_cache)
+
+            # Essentially free
+            if include_features:
+                base_feature_map = self.solve_base_feature_map(
+                    model_name, dependency_cache=root_cache)
+                model_scores[model_name]['features'] = base_feature_map
+
+        return model_scores
+
+    def solve_features(self, model_name, dependency_cache=None):
+        """
+        """
+        features = self[model_name].features
+        return list(self.extractor.solve(features, cache=dependency_cache))
+
+    def solve_base_feature_map(self, model_name, dependency_cache=None):
+        """
+        """
+        features = list(trim(self[model_name].features))
+        feature_values = self.extractor.solve(features, cache=dependency_cache)
+        return {str(f): v
+                for f, v in zip(features, feature_values)}
+
+    def process_score(self, model_name, dependency_cache=None):
+        """
+        """
+        version = self[model_name].version
+
+        start = time.time()
+        feature_values = list(self.solve_features(
+            model_name, dependency_cache))
+        logger.debug("Extracted features for {0}:{1}:{2} in {3} secs"
+                     .format(self.name, model_name, version,
+                             time.time() - start))
+
+        start = time.time()
+        score = self[model_name].score(feature_values)
+        logger.debug("Scored features for {0}:{1}:{2} in {3} secs"
+                     .format(self.name, model_name, version,
+                             time.time() - start))
+
+        return score
+
+    def _generate_root_datasources(self, model_names):
+        for model_name in model_names:
+            for dependency in dependencies.dig(self.model_features(model_name)):
+                if isinstance(dependency, Datasource):
```

Comment:
As the function name implies, we'd like to have the root Datasources.  There are other items that can be found at the root of a dependency tree -- like Constant.